exporter/azuremonitorexporter/factory.go

```diff
@@ -0,0 +1,107 @@
+	"github.com/open-telemetry/opentelemetry-collector/exporter"
+	"go.uber.org/zap"
+)
+
+const (
+	// The value of "type" key in configuration.
+	typeStr = "azuremonitor"
+)
+
+var (
+	errUnexpectedConfigurationType = errors.New("failed to cast configuration to Azure Monitor Config")
+)
+
+// Factory for Azure Monitor exporter.
+// Implements the interface from github.com/open-telemetry/opentelemetry-collector/exporter/factory.go
+type Factory struct {
+	TransportChannel transportChannel
+}
+
+// Type gets the type of the Exporter config created by this factory.
+func (f *Factory) Type() string {
+	return typeStr
+}
+
+// CreateDefaultConfig creates the default configuration for exporter.
+func (f *Factory) CreateDefaultConfig() configmodels.Exporter {
+
+	return &Config{
+		ExporterSettings: configmodels.ExporterSettings{
+			TypeVal: typeStr,
+			NameVal: typeStr,
+		},
+		Endpoint:                  "https://dc.services.visualstudio.com/v2/track",
+		MaxBatchSize:              1024,
+		MaxBatchIntervalInSeconds: int(time.Duration(10) * time.Second),
+	}
+}
+
+// CreateTraceExporter creates a trace exporter based on this config.
+func (f *Factory) CreateTraceExporter(logger *zap.Logger, config configmodels.Exporter) (exporter.TraceExporter, error) {
+	exporterConfig, ok := config.(*Config)
+
+	if !ok {
+		return nil, errUnexpectedConfigurationType
+	}
+
+	transportChannel := f.getTransportChannel(exporterConfig, logger)
+	return newTraceExporter(exporterConfig, transportChannel, logger)
+}
+
+// CreateMetricsExporter creates a metrics exporter based on this config.
+func (f *Factory) CreateMetricsExporter(
+	logger *zap.Logger,
+	cfg configmodels.Exporter,
+) (exporter.MetricsExporter, error) {
+	return nil, configerror.ErrDataTypeIsNotSupported
+}
+
+// Configures the transport channel.
+// This method is not thread-safe
+func (f *Factory) getTransportChannel(exporterConfig *Config, logger *zap.Logger) transportChannel {
+
+	// The default transport channel uses the default send mechanism from the AppInsights telemetry client.
+	// This default channel handles batching, appropriate retries, and is backed by memory.
+	if f.TransportChannel == nil {
+		telemetryConfiguration := appinsights.NewTelemetryConfiguration(exporterConfig.InstrumentationKey)
+		telemetryConfiguration.EndpointUrl = exporterConfig.Endpoint
+		telemetryConfiguration.MaxBatchSize = exporterConfig.MaxBatchSize
+		telemetryConfiguration.MaxBatchInterval = time.Duration(exporterConfig.MaxBatchIntervalInSeconds) * time.Second
+		telemetryClient := appinsights.NewTelemetryClientFromConfig(telemetryConfiguration)
+
+		f.TransportChannel = telemetryClient.Channel()
+
+		// Don't even bother enabling the AppInsights diagnostics listener unless debug logging is enabled
+		if checkedEntry := logger.Check(zap.DebugLevel, ""); checkedEntry != nil {
+			appinsights.NewDiagnosticsMessageListener(func(msg string) error {
+				logger.Debug(msg)
```

Comment:
Q.: does `msg` already comes with some prefix that makes clear that is coming from appinsights?

Reply:
The debug messages come out looking like this:

{"level":"debug","ts":1575390241.9811702,"caller":"azuremonitorexporter/factory.go:100","msg":"--------- Transmitting 4 items ---------"}
{"level":"debug","ts":1575390241.9812257,"caller":"azuremonitorexporter/factory.go:100","msg":"--------- Transmitting 4 items ---------"}
{"level":"debug","ts":1575390242.4847004,"caller":"azuremonitorexporter/factory.go:100","msg":"Telemetry transmitted in 503.295ms"}
{"level":"debug","ts":1575390242.484751,"caller":"azuremonitorexporter/factory.go:100","msg":"Response: 200"}
{"level":"debug","ts":1575390242.4847667,"caller":"azuremonitorexporter/factory.go:100","msg":"Items accepted/received: 4/4"}
{"level":"debug","ts":1575390242.5504136,"caller":"azuremonitorexporter/factory.go:100","msg":"Telemetry transmitted in 569.1559ms"}
{"level":"debug","ts":1575390242.550459,"caller":"azuremonitorexporter/factory.go:100","msg":"Response: 200"}
{"level":"debug","ts":1575390242.55047,"caller":"azuremonitorexporter/factory.go:100","msg":"Items accepted/received: 4/4"}